integrate Crypto{Engine,Socket} into fnet

havardpe opened the pull request:
@vekterli, @arnej27959  please review
@bjorncs FYI

havardpe committed:
integrate Crypto{Engine,Socket} into fnet

havardpe committed:
make sure we flush even if no new data needs writing

vekterli commented on vespalib/src/vespa/vespalib/net/crypto_engine.cpp:
```diff
@@ -155,10 +155,29 @@ class XorCryptoSocket : public CryptoSocket
+CryptoEngine::SP
+CryptoEngine::get_default()
+{
+    std::lock_guard<std::mutex> guard(_shared_lock);
```
Can simplify to `std::lock_guard guard(_shared_lock)` due to C++17 auto template parameter type inference 🦄✨

havardpe replied:
oooh nice; will start to use this feature

vekterli reviewed (approved):
:+1: LGTM! Only 1 optional micro-nitpick and a data flow question re: socket bork

havardpe committed:
fixes based on feedback

- do not check broken flag when doing flush
- use auto-detection of guard template parameter

havardpe commented on fnet/src/vespa/fnet/connection.cpp:
```diff
@@ -342,6 +386,17 @@ FNET_Connection::Write(bool direct)
+        ++my_write_work;
+    }
+
+    if ((res >= 0) && !broken) { // flush output pipeline
```
removing broken from the check since all errors are tracked as negative res at this point

havardpe committed:
drain input pipeline after crypto handshake

arnej27959 reviewed (approved):
looks good to me
